File: library/src/main/java/com/thesurix/gesturerecycler/transactions/AddTransaction.java
package com.thesurix.gesturerecycler.transactions;


import com.thesurix.gesturerecycler.GestureAdapter;
import com.thesurix.gesturerecycler.GestureViewHolder;

/**
 * @author thesurix
 */
public class AddTransaction<T> implements AdapterTransaction {

    private final GestureAdapter<T, ? extends GestureViewHolder> mAdapter;
    private final T mItem;

    public AddTransaction(final GestureAdapter<T, ? extends GestureViewHolder> adapter, final T item) {
        mAdapter = adapter;
        mItem = item;
    }

    @Override
    public boolean perform() {
        final boolean success = mAdapter.getData().add(mItem);
        if (success) {
            mAdapter.notifyItemInserted(mAdapter.getItemCount());
        }
        return success;
    }

    @Override
    public boolean revert() {
        final int dataSize = mAdapter.getItemCount();
        final T item = mAdapter.getData().remove(dataSize - 1);
        final boolean success = item != null;
        if (success) {
            mAdapter.notifyItemRemoved(dataSize);
        }
        return success;
    }
}

File: library/src/main/java/com/thesurix/gesturerecycler/transactions/AddTransaction.kt
package com.thesurix.gesturerecycler.transactions


/**
 * @author thesurix
 */
class AddTransaction<T>(private val item: T,
                        private val headerEnabled: Boolean) : Transaction<T> {

    override fun perform(transactional: Transactional<T>): Boolean {
        return with(transactional.data) {
            val success = add(item)
            if (success) {
                transactional.notifyInserted(if (headerEnabled) size else size - 1)
            }
            success
        }
    }

    override fun revert(transactional: Transactional<T>): Boolean {
        return with(transactional.data) {
            val item = removeAt(size - 1)
            item?.let {
                transactional.notifyRemoved(if (headerEnabled) size + 1 else size)
                true
            } ?: false
        }
    }
}
